### ISSUE TITLE
503 ERROR due lambda timeout after tokens are fetched

### ISSUE THREAD

**Issue by mishabruml** [awslabs/cognito-at-edge#86]
#### What happened:
My cloudfront dist redirects me to the cognito login UI and I successfully auth. After that, I get directed to the cloudfront 503 error page rather than my s3 static content:

```
503 ERROR
The request could not be satisfied.
The Lambda function associated with the CloudFront distribution is invalid or doesn't have the required permissions. We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error. Try again later, or contact the app or website owner.
If you provide content to customers through CloudFront, you can find steps to troubleshoot and help prevent this error by reviewing the CloudFront documentation.

Generated by cloudfront (CloudFront)
Request ID: BO_FjHzTRlnA-HWDVRWznjZbcpOBBMjEAT2mwvZUxje6BITPn2bbJg==
```

The logs in my lambda@edge appear to show the initial invocation successfully recirecting the user to the UI page:

<details>
<summary>logs</summary>

```
{
    "level": 20,
    "time": 1705676579380,
    "msg": "Handling Lambda@Edge event",
    "event": {
        "Records": [ ....big blob redacted here ]
    }
}
{
    "level": 20,
    "time": 1705676579400,
    "msg": "Cookies weren't present in the request"
}
{
    "level": 20,
    "time": 1705676579400,
    "msg": "User isn't authenticated: Error: Cookies weren't present in the request"
}
{
    "level": 20,
    "time": 1705676579400,
    "msg": "Redirecting user to Cognito User Pool URL https://***********"
}
```
</details>

Then after entering my login credentials, after 5s I am presented with the 503 screen in the browser. The lambda appears to time out after fetching the tokens.

<details>

<summary>logs</summary>

```
{
    "level": 20,
    "time": 1705676947082,
    "msg": "Handling Lambda@Edge event",
    "event": {
        "Records": [...redacted blob]
    }
}
{
    "level": 20,
    "time": 1705676947102,
    "msg": "Cookies weren't present in the request"
}
{
    "level": 20,
    "time": 1705676947102,
    "msg": "User isn't authenticated: Error: Cookies weren't present in the request"
}
{
    "level": 20,
    "time": 1705676947102,
    "msg": "Fetching tokens from grant code...",
    "request": {
        "url": "https://****.auth.****.amazoncognito.com/oauth2/token",
        "method": "POST",
        "headers": {
            "Content-Type": "application/x-www-form-urlencoded"
        },
        "data": "client_id=******&code=******&grant_type=authorization_code&redirect_uri=******.cloudfront.net"
    },
    "code": "******"
}
{
    "level": 20,
    "time": 1705676949484,
    "msg": "Fetched tokens",
    "tokens": {
        "id_token": "*****",
        "access_token": "****",
        "refresh_token": "****",
        "expires_in": 28800,
        "token_type": "Bearer"
    }
}
```
</details>

**Comment by mishabruml** [comment]
I did some debugging afger forking the repo and adding logs everywhere, and found that the lambda was timing out here https://github.com/awslabs/cognito-at-edge/blob/e8c7e305b2fe87d0fbc47bcde560b4526a0235d2/src/index.ts#L279 verifying the JWT token. Others appeared to have had this issue https://github.com/awslabs/aws-jwt-verify/issues/72 https://github.com/awslabs/aws-jwt-verify/issues/133 so I have made a PR #88 to address this. Changing the timeout to 5000ms fixed the issue for me.

**Comment by ckifer** [comment]
+1 I have the same issue

**Comment by ckifer** [comment]
Now I'm experiencing this after resolving the other issue. Still +1

> 2024-01-25T02:28:37.114Z 9c52cf55-4769-454a-8764-329342c781c2 Task timed out after 3.06 seconds

**Comment by lenfree** [comment]
Not sure if these would address the 5 second time out issue:

1.  increase the timeout to 5 seconds
2.  enable the `AWS_NODEJS_CONNECTION_REUSE_ENABLED` 
```node
    process.env['AWS_NODEJS_CONNECTION_REUSE_ENABLED'] = '1';
```

**Comment by ckifer** [comment]
Increasing the timeout worked for me... for now. Lambda@edge doesn't support timeouts > 5 seconds though. Also doesn't support env bars

**Comment by lenfree** [comment]
> ```js
>   process.env['AWS_NODEJS_CONNECTION_REUSE_ENABLED'] = '1';
> ```

That's right, it does not support Lambda environment variables but you can manually set it with below if I am not mistaken. Cold start might also contribute to the issue.

```
process.env['AWS_NODEJS_CONNECTION_REUSE_ENABLED'] = '1';
```

**Comment by ckifer** [comment]
Can give it a shot. Though I think that it's mostly because of region distance. An edge function in Singapore is making a call to Cognito in Oregon

**Comment by manu-remsense** [comment]
Hey guys, I was having the same issue, and thankfully I was able to come across this issue and resolve the problem with what @mishabruml has suggested, thanks 🙏 .

**Comment by ckifer** [comment]
Sweet, thanks for the info!

**Comment by manu-remsense** [comment]
Actually it didn't help, I tested it wrong. I understand the issue on my end a bit better, and turns out I have the same problem as you @ckifer. When a token is expired it has to verify the token, then verify the refresh token and then fetch the new token - process seems to longer than 5s time limit when cognito and cloudfront location are far apart. I still have the timeout increased for the `SimpleJsonFetcher` in place but it hasn't really resolved it for me. Did you manage to find something that fixes it @ckifer ? 
Thanks

**Comment by ckifer** [comment]
No resolution here no. Luckily I don't have too many users that are far away, but still would like to find a better solution.

**Comment by lenfree** [comment]
I had the same issue too and so far [this](https://github.com/awslabs/cognito-at-edge/issues/86#issuecomment-1920452251) worked for me, or maybe I just haven't tested it properly.

**Comment by manu-remsense** [comment]
> I had the same issue too and so far [this](https://github.com/awslabs/cognito-at-edge/issues/86#issuecomment-1920452251) worked for me, or maybe I just haven't tested it properly.

@lenfree I've put that line `process.env['AWS_NODEJS_CONNECTION_REUSE_ENABLED'] = '1';` outside of the handler, and it didn't work. Or did I not do it correctly? Thanks

**Comment by manu-remsense** [comment]
Anyway, as a band-aid solution we decided to forgo the use of refresh tokens when a user is accessing outside of our region (australia)

 -                if (tokens.refreshToken) {
+                if (tokens.refreshToken && process.env.AWS_REGION === 'ap-southeast-2') {

**Comment by ckifer** [comment]
That's actually a solid workaround, nice!

**Comment by lenfree** [comment]
> > I had the same issue too and so far [this](https://github.com/awslabs/cognito-at-edge/issues/86#issuecomment-1920452251) worked for me, or maybe I just haven't tested it properly.
> 
> @lenfree I've put that line `process.env['AWS_NODEJS_CONNECTION_REUSE_ENABLED'] = '1';` outside of the handler, and it didn't seem to improve the situation. Or did I not do it correctly? Thanks

I think it needs to be inside the handler, at least that's how I configured it.

**Comment by ckifer** [comment]
>  if (tokens.refreshToken && process.env.AWS_REGION === 'ap-southeast-2')

@manu-remsense  Is this you modifying the cognito-at-edge code? Wish there was a config item for that...

**Comment by manu-remsense** [comment]
> > > I had the same issue too and so far [this](https://github.com/awslabs/cognito-at-edge/issues/86#issuecomment-1920452251) worked for me, or maybe I just haven't tested it properly.
> > 
> > 
> > @lenfree I've put that line `process.env['AWS_NODEJS_CONNECTION_REUSE_ENABLED'] = '1';` outside of the handler, and it didn't seem to improve the situation. Or did I not do it correctly? Thanks
> 
> I think it needs to be inside the handler, at least that's how I configured it.

Ohh I see, I should give it a try 🤔

@ckifer yes that's the only way as far as I know, I can make a PR later when I have time on the weekend (but not sure how useful it would be to people).

**Comment by BredoGen** [comment]
[aws-jwt-verify](https://github.com/awslabs/aws-jwt-verify) v5.0.0 raised the default timeout to 3000ms.

**Comment by ciaochen-amazon** [comment]
I have the same issue. After I upgrade from nodeJS 14 to nodeJS20. It start to throw 503 error, and suggesting time out from lambda. 

I suspect it's because the memory issue, because I did a test run against the lambda with 128MB memory , and it shows times out with all memory used and thus killed the lambda. And then I made a second test run against the lambda with 500MB memory and this time it works, at last it successfully initialized the lambda.
